sklearn/utils/_indexing.py

```diff
@@ -264,9 +269,12 @@ def _safe_indexing(X, indices, *, axis=0):
     if hasattr(X, "iloc"):
         # TODO: we should probably use _is_pandas_df_or_series(X) instead but this
         # would require updating some tests such as test_train_test_split_mock_pandas.
+        # TODO: Should also work with _narwhals_indexing, but
+        # test_safe_indexing_pandas_no_settingwithcopy_warning
+        # does not pass.
         return _pandas_indexing(X, indices, indices_dtype, axis=axis)
-    elif _is_polars_df_or_series(X):
-        return _polars_indexing(X, indices, indices_dtype, axis=axis)
+    elif nw.dependencies.is_into_dataframe(X) or nw.dependencies.is_into_series(X):
+        return _narwhals_indexing(X, indices, indices_dtype, axis=axis)
```

Comment:
This whole section should read
```python
    if nw.dependencies.is_into_dataframe(X) or nw.dependencies.is_into_series(X):
        return _narwhals_indexing(X, indices, indices_dtype, axis=axis)
```
and the pandas/iloc path should be removed. But currently, it does not pass the test `test_safe_indexing_pandas_no_settingwithcopy_warning`.

@MarcoGorelli You might have additional insights.

Reply:
thanks for the ping, i'll take a look next week!

Reply:
To be honest I don't understand this test. It checks that pandas makes a copy, but only when indexing with `[0, 1]`. On the `main` branch, changing that to `slice(0, 2)` is enough to make the test fail

I'll open a separate issue to discuss

Narwhals objects (DataFrame, Series) are immutable so this isn't something you should need to worry about with Narwhals anyway

Reply:
Here you go: https://github.com/scikit-learn/scikit-learn/issues/31290

Anyone in particular we should tag in it?